staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/fairqueuing.go

```diff
@@ -0,0 +1,621 @@
+		total += queue.RequestsExecuting
+	}
+	return total
+}
+
+// ChooseQueueIdx uses shuffle sharding to select an queue index
+// using a 'hashValue'.  The 'hashValue' derives a hand from a set range of
+// indexes (range 'desiredNumQueues') and returns the queue with the least queued packets
+// from a dealt hand (of size 'handSize')
+func (qs *queueSetImpl) ChooseQueueIdx(hashValue uint64, handSize int) int {
+	// TODO(aaron-prindle) currently a lock is held for this in a larger anonymous function
+	// verify that makes sense...
+
+	bestQueueIdx := -1
+	bestQueueLen := int(math.MaxInt32)
+	// desiredNumQueues is used here instead of numQueues to omit quiesce queues
+	shufflesharding.ShuffleAndDeal(hashValue, qs.desiredNumQueues, handSize, func(queueIdx int) {
+		thisLen := len(qs.queues[queueIdx].Requests)
+		if thisLen < bestQueueLen {
+			bestQueueIdx, bestQueueLen = queueIdx, thisLen
+		}
+	})
+	return bestQueueIdx
+}
+
+// rejectOrEnqueue rejects or enqueues the newly arrived request if
+// resource criteria isn't met
+func (qs *queueSetImpl) rejectOrEnqueue(packet *Request) bool {
+	queue := packet.Queue
+	curQueueLength := len(queue.Requests)
+	// rejects the newly arrived request if resource criteria not met
+	if qs.GetRequestsExecuting() >= qs.concurrencyLimit &&
+		curQueueLength >= qs.queueLengthLimit {
+		return false
+	}
+
+	qs.enqueue(packet)
+	return true
+}
+
+// enqueues a packet into an queueSetImpl
+func (qs *queueSetImpl) enqueue(packet *Request) {
+
+	queue := packet.Queue
+	queue.Enqueue(packet)
+	qs.updateQueueVirStartTime(packet, queue)
+	qs.numRequestsEnqueued++
+
+	metrics.UpdateFlowControlRequestsInQueue(qs.name, qs.numRequestsEnqueued)
+}
+
+// Enqueue enqueues a packet directly into an queueSetImpl w/ no restriction
+func (qs *queueSetImpl) Enqueue(packet *Request) bool {
+	qs.LockAndSyncTime()
+	defer qs.lock.Unlock()
+
+	qs.enqueue(packet)
+	return true
+}
+
+// synctime is used to sync the time of the queueSetImpl by looking at the elapsed
+// time since the last sync and this value based on the 'virtualtime ratio'
+// which scales inversely to the # of active flows
+func (qs *queueSetImpl) synctime() {
+	realNow := qs.clk.Now()
+	timesincelast := realNow.Sub(qs.lastRealTime).Seconds()
+	qs.lastRealTime = realNow
+	qs.vt += timesincelast * qs.getvirtualtimeratio()
+}
+
+func (qs *queueSetImpl) getvirtualtimeratio() float64 {
+	NEQ := 0
+	reqs := 0
+	for _, queue := range qs.queues {
+		reqs += queue.RequestsExecuting
+		// It might be best to delete this line. If everything is working
+		//  correctly, there will be no waiting packets if reqs < C on current
+		//  line 85; if something is going wrong, it is more accurate to say
+		// that virtual time advanced due to the requests actually executing.
+
+		// reqs += len(queue.Requests)
+		if len(queue.Requests) > 0 || queue.RequestsExecuting > 0 {
+			NEQ++
+		}
+	}
+	// no active flows, virtual time does not advance (also avoid div by 0)
+	if NEQ == 0 {
```

Comment:
I think you meant nonEmptyQueue - better use full name

Reply:
The full description is "number of non-empty queues".